tasks/vagrant.py

```diff
@@ -5,6 +5,38 @@
 from .common import PopenTask, PopenException, FallibleTask, TaskException
 
 
+def with_vagrant(func):
+    def wrapper(self, *args, **kwargs):
+        try:
+            __setup_provision(self)
+        except TaskException as exc:
+            logging.critical('vagrant or provisioning failed')
+            raise exc
+        else:
+            func(self, *args, **kwargs)
+        finally:
+            if not self.no_destroy:
+                self.execute_subtask(
+                    VagrantCleanup(raise_on_err=False))
+
+    return wrapper
+
+
+def __setup_provision(self):
```

Comment:
Well, I though to change `self` to `instance` (or something like that, but I think it would even more confuse.

I think that, by the reason that use `self` outside of a class method is weird, we should keep the `self` in `__setup_provision`. IMHO it's more explicit.